src/qibolab/compilers/compiler.py

```diff
@@ -0,0 +1,143 @@
+        virtual_z_phases (dict): Dictionary mapping qubits to virtual Z-phases induced by the gate.
+
+    See ``qibolab.compilers.default`` for an example of a compiler implementation.
+    """
+
+    rules: dict = field(default_factory=dict)
+    """Map from gates to compilation rules."""
+
+    @classmethod
+    def default(cls):
+        return cls(
+            {
+                gates.I: identity_rule,
+                gates.Z: z_rule,
+                gates.RZ: rz_rule,
+                gates.U3: u3_rule,
+                gates.CZ: cz_rule,
+                gates.M: measurement_rule,
+            }
+        )
+
+    def __setitem__(self, key, rule):
+        """Sets a new rule to the compiler.
+
+        If a rule already exists for the gate, it will be overwritten.
+        """
+        self.rules[key] = rule
+
+    def __getitem__(self, item):
+        """Get an existing rule for a given gate."""
+        if item not in self.rules:
+            raise_error(KeyError, f"Compiler rule not available for {item}.")
+        return self.rules[item]
+
+    def __delitem__(self, item):
+        """Remove rule for the given gate."""
+        if item not in self.rules:
+            raise_error(KeyError, f"Cannot remove {item} from compiler because it does not exist.")
+        del self.rules[item]
+
+    def register(self, gate_cls):
+        """Decorator for registering a function as a rule in the compiler.
+
+        Using this decorator is optional. Alternatively the user can set the rules directly
+        via ``__setitem__`.
+
+        Args:
+            gate_cls: Qibo gate object that the rule will be assigned to.
+        """
+
+        def inner(func):
+            self[gate_cls] = func
+            return func
+
+        return inner
+
+    def compile(self, circuit, platform):
+        """Transforms a circuit to pulse sequence.
+
+        Args:
+            circuit (qibo.models.Circuit): Qibo circuit that respects the platform's
+                connectivity and native gates.
+            platform (qibolab.platforms.abstract.AbstractPlatform): Platform used
+                to load the native pulse representations.
+
+        Returns:
+            sequence (qibolab.pulses.PulseSequence): Pulse sequence that implements the circuit.
+            measurement_map (dict): Map from each measurement gate to the sequence of  readout pulses
+                implementing it.
+        """
+        sequence = PulseSequence()
+        # FIXME: This will not work with qubits that have string names
+        # TODO: Implement a mapping between circuit qubit ids and platform ``Qubit``s
+        virtual_z_phases = defaultdict(int)
+
+        measurement_map = {}
+        # keep track of gates that were already added to avoid adding them twice
+        already_processed = set()
+        # process circuit gates
+        for moment in circuit.queue.moments:
+            moment_start = sequence.finish
+            for gate in moment:
```

Comment:
Indeed, it is possible to get rid of `already_processed` using the `set`.

However, I am not sure if I can get rid of the double iteration, because I think `moment_start` needs to be calculated before starting to process each moment and be used for all gates in that moment.

I am not sure how I would properly convert the `moment_start` calculation from
```py
for moment in circuit.queue.moments:
    moment_start = sequence.finish
    for gate in set(moment):
```
to a single loop.

Reply:
For the time being, apply the easy one (drop `already_processed`) and let's keep the complex (denesting loops) for a later stage.

The idea is just to make the code as simple as possible, and in this perspective there are also greater benefits in `already_processed` removal (since this will decouple more loops), rather than folding in a single one, if there isn't an intuitive way of doing it.

Reply:
> For the time being, apply the easy one (drop `already_processed`) and let's keep the complex (denesting loops) for a later stage.

Already done.